src/main/java/com/hivemq/extensions/log/mqtt/message/config/ExtensionConfigReader.java

```diff
@@ -0,0 +1,177 @@
+import static com.hivemq.extensions.log.mqtt.message.config.ExtensionConfigProperties.CONNACK_SEND;
+import static com.hivemq.extensions.log.mqtt.message.config.ExtensionConfigProperties.FALSE;
+import static com.hivemq.extensions.log.mqtt.message.config.ExtensionConfigProperties.PAYLOAD;
+import static com.hivemq.extensions.log.mqtt.message.config.ExtensionConfigProperties.PING_REQUEST_RECEIVED;
+import static com.hivemq.extensions.log.mqtt.message.config.ExtensionConfigProperties.PING_RESPONSE_SEND;
+import static com.hivemq.extensions.log.mqtt.message.config.ExtensionConfigProperties.PUBACK_RECEIVED;
+import static com.hivemq.extensions.log.mqtt.message.config.ExtensionConfigProperties.PUBACK_SEND;
+import static com.hivemq.extensions.log.mqtt.message.config.ExtensionConfigProperties.PUBCOMP_RECEIVED;
+import static com.hivemq.extensions.log.mqtt.message.config.ExtensionConfigProperties.PUBCOMP_SEND;
+import static com.hivemq.extensions.log.mqtt.message.config.ExtensionConfigProperties.PUBLISH_RECEIVED;
+import static com.hivemq.extensions.log.mqtt.message.config.ExtensionConfigProperties.PUBLISH_SEND;
+import static com.hivemq.extensions.log.mqtt.message.config.ExtensionConfigProperties.PUBREC_RECEIVED;
+import static com.hivemq.extensions.log.mqtt.message.config.ExtensionConfigProperties.PUBREC_SEND;
+import static com.hivemq.extensions.log.mqtt.message.config.ExtensionConfigProperties.PUBREL_RECEIVED;
+import static com.hivemq.extensions.log.mqtt.message.config.ExtensionConfigProperties.PUBREL_SEND;
+import static com.hivemq.extensions.log.mqtt.message.config.ExtensionConfigProperties.SUBACK_SEND;
+import static com.hivemq.extensions.log.mqtt.message.config.ExtensionConfigProperties.SUBSCRIBE_RECEIVED;
+import static com.hivemq.extensions.log.mqtt.message.config.ExtensionConfigProperties.TRUE;
+import static com.hivemq.extensions.log.mqtt.message.config.ExtensionConfigProperties.UNSUBACK_SEND;
+import static com.hivemq.extensions.log.mqtt.message.config.ExtensionConfigProperties.UNSUBSCRIBE_RECEIVED;
+import static com.hivemq.extensions.log.mqtt.message.config.ExtensionConfigProperties.VERBOSE;
+
+/**
+ * @since 1.2.0
+ */
+public class ExtensionConfigReader {
+
+    private static final @NotNull Logger LOG = LoggerFactory.getLogger(ExtensionConfigReader.class);
+
+    /**
+     * Reads config.xml from /conf directory if present, otherwise reads mqttMessageLog.properties from extension home
+     * directory. Uses the defaults if neither is present/readable.
+     *
+     * @param extensionHomeFolder home folder of the extension
+     * @return user defined extension configuration or default configuration
+     */
+    public static @NotNull ExtensionConfig read(final @NotNull File extensionHomeFolder) {
+        final File configXmlFile = extensionHomeFolder.toPath().resolve(EXTENSION_CONFIG_XML_LOCATION).toFile();
+
+        if (!configXmlFile.exists()) {
```

Comment:
The code feels a bit disconnected from the logical ordering here.

1. we want to read the xml config, if exists and valid -> use
2. we want to read the properties file, if exists and valid -> use
3. in other cases use default configuration

Why don't we create a default ExtensionConfig that is used to set the defaults in both PropertiesConfig and XmlConfig or set if no configs are existing. Personally, I would not use an inheritance tree but pass the default ExtensionConfig object.

This would also prevent the "need to keep in sync"

Reply:
Or different approach:

We have a single ExtensionConfig class that is almost equal to the current ExtensionConfigXml except it also has setters.

There, the defaults are set with the field initializers. In case of xml initialization, the annotations are used. In case of properties initialization, a method is used that has a switch statement (or equal) to set the no-arg constructed ExtensionConfig with the setters.

Technically, ExtensionConfigProperties and ExtensionConfigXml do not provide different implementations of the ExtensionConfig.

Reply:
I wanted to make both configs as separate as possible to it's easy to remove the legacy config, while also making the least amount of modifications possible to the codebase, as this should be a straightforward card and not a refactor.

Reply:
Will create a follow-up card.